// @ljcl/storybook-addon-cssprops/dist/cjs/components/CssPropsTable/CssPropRow.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CssPropRow = void 0;
var React = require("react");
var theming_1 = require("@storybook/theming");
var CssPropControl_1 = require("./CssPropControl");
var shared_1 = require("../typography/shared");
var Name = theming_1.styled.code(function (_a) {
    var theme = _a.theme;
    return (__assign({ fontWeight: "bold", textAlign: "left" }, (0, shared_1.codeCommon)({ theme: theme })));
});
var Description = theming_1.styled.div(function (_a) {
    var theme = _a.theme;
    return ({
        "&&": {
            p: {
                margin: "0 0 10px 0",
            },
            a: {
                color: theme.color.secondary,
            },
        },
        code: (0, shared_1.codeCommon)({ theme: theme }),
        "& code": {
            margin: 0,
            display: "inline-block",
        },
    });
});
var StyledTd = theming_1.styled.td(function () { return ({
    paddingLeft: "20px !important",
}); });
var CssPropRow = function (props) {
    var row = props.row, updateStorage = props.updateStorage;
    var name = row.name, description = row.description;
    var hasDescription = description != null && description !== "";
    return (React.createElement("tr", null,
        React.createElement(StyledTd, null,
            React.createElement(Name, null,
                "--",
                name)),
        React.createElement("td", null, hasDescription && React.createElement(Description, null, description)),
        React.createElement("td", null, updateStorage ? (React.createElement(CssPropControl_1.CssPropControl, __assign({}, props))) : null)));
};
exports.CssPropRow = CssPropRow;
