// @ljcl/storybook-addon-cssprops/dist/esm/components/CssPropsTable/CssPropRow.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as React from "react";
import { styled } from "@storybook/theming";
import { CssPropControl } from "./CssPropControl";
import { codeCommon } from "../typography/shared";
var Name = styled.code(function (_a) {
    var theme = _a.theme;
    return (__assign({ fontWeight: "bold", textAlign: "left" }, codeCommon({ theme: theme })));
});
var Description = styled.div(function (_a) {
    var theme = _a.theme;
    return ({
        "&&": {
            p: {
                margin: "0 0 10px 0",
            },
            a: {
                color: theme.color.secondary,
            },
        },
        code: codeCommon({ theme: theme }),
        "& code": {
            margin: 0,
            display: "inline-block",
        },
    });
});
var StyledTd = styled.td(function () { return ({
    paddingLeft: "20px !important",
}); });
export var CssPropRow = function (props) {
    var row = props.row, updateStorage = props.updateStorage;
    var name = row.name, description = row.description;
    var hasDescription = description != null && description !== "";
    return (React.createElement("tr", null,
        React.createElement(StyledTd, null,
            React.createElement(Name, null,
                "--",
                name)),
        React.createElement("td", null, hasDescription && React.createElement(Description, null, description)),
        React.createElement("td", null, updateStorage ? (React.createElement(CssPropControl, __assign({}, props))) : null)));
};
