// @ljcl/storybook-addon-cssprops/dist/cjs/components/typography/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withMargin = exports.withReset = exports.headerCommon = exports.codeCommon = void 0;
var polished_1 = require("polished");
var codeCommon = function (_a) {
    var theme = _a.theme;
    return ({
        lineHeight: 1,
        margin: '0 2px',
        padding: '3px 5px',
        whiteSpace: 'nowrap',
        borderRadius: 3,
        fontSize: theme.typography.size.s2 - 1,
        border: theme.base === 'light'
            ? "1px solid ".concat(theme.color.mediumlight)
            : "1px solid ".concat(theme.color.darker),
        color: theme.base === 'light'
            ? (0, polished_1.transparentize)(0.1, theme.color.defaultText)
            : (0, polished_1.transparentize)(0.3, theme.color.defaultText),
        backgroundColor: theme.base === 'light' ? theme.color.lighter : theme.color.border,
    });
};
exports.codeCommon = codeCommon;
var headerCommon = function (_a) {
    var theme = _a.theme;
    return ({
        margin: '20px 0 8px',
        padding: 0,
        cursor: 'text',
        position: 'relative',
        color: theme.color.defaultText,
        '&:first-of-type': {
            marginTop: 0,
            paddingTop: 0,
        },
        '&:hover a.anchor': {
            textDecoration: 'none',
        },
        '& tt, & code': {
            fontSize: 'inherit',
        },
    });
};
exports.headerCommon = headerCommon;
var withReset = function (_a) {
    var theme = _a.theme;
    return ({
        fontFamily: theme.typography.fonts.base,
        fontSize: theme.typography.size.s3,
        margin: 0,
        WebkitFontSmoothing: 'antialiased',
        MozOsxFontSmoothing: 'grayscale',
        WebkitTapHighlightColor: 'rgba(0, 0, 0, 0)',
        WebkitOverflowScrolling: 'touch',
    });
};
exports.withReset = withReset;
exports.withMargin = {
    margin: '16px 0',
};
